src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs

```diff
@@ -0,0 +1,360 @@
+    ///     <description><see cref="ImmutableDictionary{TKey, TValue}"/> Complexity</description>
+    ///     <description>Comments</description>
+    ///   </item>
+    ///   <item>
+    ///     <description>Item</description>
+    ///     <description>O(1)</description>
+    ///     <description>O(log n)</description>
+    ///     <description>Directly index into the underlying segmented dictionary</description>
+    ///   </item>
+    ///   <item>
+    ///     <description>Add()</description>
+    ///     <description>O(n)</description>
+    ///     <description>O(log n)</description>
+    ///     <description>Requires creating a new segmented dictionary</description>
+    ///   </item>
+    /// </list>
+    /// 
+    /// <para>This type is backed by segmented arrays to avoid using the Large Object Heap without impacting algorithmic
+    /// complexity.</para>
+    /// </remarks>
+    /// <typeparam name="TKey">The type of the keys in the dictionary.</typeparam>
+    /// <typeparam name="TValue">The type of the values in the dictionary.</typeparam>
+    /// <devremarks>
+    /// <para>This type has a documented contract of being exactly one reference-type field in size. Our own
+    /// <see cref="RoslynImmutableInterlocked"/> class depends on it, as well as others externally.</para>
+    ///
+    /// <para><strong>IMPORTANT NOTICE FOR MAINTAINERS AND REVIEWERS:</strong></para>
+    ///
+    /// <para>This type should be thread-safe. As a struct, it cannot protect its own fields from being changed from one
+    /// thread while its members are executing on other threads because structs can change <em>in place</em> simply by
+    /// reassigning the field containing this struct. Therefore it is extremely important that <strong>⚠⚠ Every member
+    /// should only dereference <c>this</c> ONCE ⚠⚠</strong>. If a member needs to reference the
+    /// <see cref="_dictionary"/> field, that counts as a dereference of <c>this</c>. Calling other instance members
+    /// (properties or methods) also counts as dereferencing <c>this</c>. Any member that needs to use <c>this</c> more
+    /// than once must instead assign <c>this</c> to a local variable and use that for the rest of the code instead.
+    /// This effectively copies the one field in the struct to a local variable so that it is insulated from other
+    /// threads.</para>
+    /// </devremarks>
+    internal readonly partial struct ImmutableSegmentedDictionary<TKey, TValue> : IImmutableDictionary<TKey, TValue>, IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>, IDictionary, IEquatable<ImmutableSegmentedDictionary<TKey, TValue>>
+        where TKey : notnull
+    {
+        public static readonly ImmutableSegmentedDictionary<TKey, TValue> Empty = new(new SegmentedDictionary<TKey, TValue>());
+
+        private readonly SegmentedDictionary<TKey, TValue> _dictionary;
+
+        private ImmutableSegmentedDictionary(SegmentedDictionary<TKey, TValue> dictionary)
```

Comment:
We could consider adding a debug flag to `SegmentedDictionary`, set here, that marks it as "frozen", and changes after assert in debug mode, to avoid potential use-after-move bugs.

Reply:
I brought over a bunch of tests from dotnet/runtime (originally for `ImmutableDictionary<TKey, TValue>`), which revealed a bunch of issues that are now fixed. I'm not sure this particular bug is likely enough to make it worth the complexity.